nan.h

```diff
@@ -2280,7 +2280,7 @@ inline void AsyncExecute (uv_work_t* req) {
   worker->Execute();
 }
 
-inline void AsyncExecuteComplete (uv_work_t* req) {
+inline void AsyncExecuteComplete(uv_work_t* req, int status = 0) {
```

Comment:
Sure we can, but for me this is the root cause of the problem - AsyncExecuteComplete can be called with 1 or with 2 arguments - not depending on the circumstances like most optional arguments, but depending on the libuv version
So this is, in a way, a variable argument function 😃

Reply:
Hehe, yes. I do understand the idea. But it's not really. I'm a bit rusty, but IIRC a function with a default argument value behaves exactly like an overloaded function. There are two valid signatures. The compiler has to choose at *compile-time*. So, how is this not ambiguous? Anyway, since your solution does not get rid of the hard `reinterpret_cast<>()` either, I don't see a real advantage of doing it this way.

It's not that I'm totally opposed here, but to merge this I need a thorough explanation how and why this works. At the very least it needs a source code comment. In two weeks nobody knows why this thing has a default argument. You have to admit, considering that it just side-steps a warning it is rather obscure...

Reply:
Here is a meta function that avoids the warning *and* gets rid of the cast. I didn't really test it, but this should do it:

````
inline void AsyncExecuteComplete (uv_work_t* req) {
  AsyncWorker* worker = static_cast<AsyncWorker*>(req->data);
  worker->WorkComplete();
  worker->Destroy();
}

template <typename CB> class UVAfterWorkAdapter;

template <>
class UVAfterWorkAdapter<void (*)(uv_work_t *)> {
  void invoke(uv_work_t * req) { AsyncExecuteComplete(req); }
};

template <>
class UVAfterWorkAdapter<void (*)(uv_work_t *, int)> {
  void invoke(uv_work_t * req, int status) { AsyncExecuteComplete(req); }
};

inline void AsyncQueueWorker (AsyncWorker* worker) {
  uv_queue_work(
      GetCurrentEventLoop()
    , &worker->request
    , AsyncExecute
    , UVAfterWorkAdapter<uv_after_work_cb>::invoke
  );
}
````

Reply:
Very good point, the cast is not needed with the optional parameter at all
No need for an intermediate template

Reply:
Ah, nice! Yes, *that* I can understand. How did I miss that? 🤔 Anyway... could you add a little comment on `AsyncExecuteComplete` explaining the default argument?